From 3cadcfd8578d6ef938244fd78faf693a1f59ee2b Mon Sep 17 00:00:00 2001
From: Jeffy Chen <jeffy.chen@rock-chips.com>
Date: Tue, 8 Jun 2021 09:09:38 +0800
Subject: [PATCH 17/42] waylandsink: Use create_immed to create dmabuf

The async creation has lots of issues when the device overloaded.

Signed-off-by: Jeffy Chen <jeffy.chen@rock-chips.com>
---
 gst-libs/gst/wayland/gstwllinuxdmabuf.c | 61 ++-----------------------
 1 file changed, 5 insertions(+), 56 deletions(-)

diff --git a/gst-libs/gst/wayland/gstwllinuxdmabuf.c b/gst-libs/gst/wayland/gstwllinuxdmabuf.c
index ac53e06..3f4c468 100644
--- a/gst-libs/gst/wayland/gstwllinuxdmabuf.c
+++ b/gst-libs/gst/wayland/gstwllinuxdmabuf.c
@@ -46,41 +46,9 @@ gst_wl_linux_dmabuf_init_once (void)
 
 typedef struct
 {
-  GMutex lock;
-  GCond cond;
   struct wl_buffer *wbuf;
 } ConstructBufferData;
 
-static void
-create_succeeded (void *data, struct zwp_linux_buffer_params_v1 *params,
-    struct wl_buffer *new_buffer)
-{
-  ConstructBufferData *d = data;
-
-  g_mutex_lock (&d->lock);
-  d->wbuf = new_buffer;
-  zwp_linux_buffer_params_v1_destroy (params);
-  g_cond_signal (&d->cond);
-  g_mutex_unlock (&d->lock);
-}
-
-static void
-create_failed (void *data, struct zwp_linux_buffer_params_v1 *params)
-{
-  ConstructBufferData *d = data;
-
-  g_mutex_lock (&d->lock);
-  d->wbuf = NULL;
-  zwp_linux_buffer_params_v1_destroy (params);
-  g_cond_signal (&d->cond);
-  g_mutex_unlock (&d->lock);
-}
-
-static const struct zwp_linux_buffer_params_v1_listener params_listener = {
-  create_succeeded,
-  create_failed
-};
-
 struct wl_buffer *
 gst_wl_linux_dmabuf_construct_wl_buffer (GstBuffer * buf,
     GstWlDisplay * display, const GstVideoInfoDmaDrm * drm_info)
@@ -95,7 +63,6 @@ gst_wl_linux_dmabuf_construct_wl_buffer (GstBuffer * buf,
   GstVideoMeta *vmeta;
   guint nplanes = 0, flags = 0;
   struct zwp_linux_buffer_params_v1 *params;
-  gint64 timeout;
   ConstructBufferData data;
 
   g_return_val_if_fail (gst_wl_display_check_format_for_dmabuf (display,
@@ -105,10 +72,6 @@ gst_wl_linux_dmabuf_construct_wl_buffer (GstBuffer * buf,
   fourcc = drm_info->drm_fourcc;
   modifier = drm_info->drm_modifier;
 
-  g_cond_init (&data.cond);
-  g_mutex_init (&data.lock);
-  g_mutex_lock (&data.lock);
-
   vmeta = gst_buffer_get_video_meta (buf);
   if (vmeta) {
     width = vmeta->width;
@@ -125,6 +88,7 @@ gst_wl_linux_dmabuf_construct_wl_buffer (GstBuffer * buf,
   } else {
     GST_ERROR_OBJECT (display, "GstVideoMeta is needed to carry DMABuf using "
         "'memory:DMABuf' caps feature.");
+    data.wbuf = NULL;
     goto out;
   }
 
@@ -167,21 +131,10 @@ gst_wl_linux_dmabuf_construct_wl_buffer (GstBuffer * buf,
     }
   }
 
-  /* Request buffer creation */
-  zwp_linux_buffer_params_v1_add_listener (params, &params_listener, &data);
-  zwp_linux_buffer_params_v1_create (params, width, height, fourcc, flags);
-
-  /* Wait for the request answer */
-  wl_display_flush (gst_wl_display_get_display (display));
-  data.wbuf = (gpointer) 0x1;
-  timeout = g_get_monotonic_time () + G_TIME_SPAN_SECOND;
-  while (data.wbuf == (gpointer) 0x1) {
-    if (!g_cond_wait_until (&data.cond, &data.lock, timeout)) {
-      GST_ERROR_OBJECT (mem->allocator, "zwp_linux_buffer_params_v1 time out");
-      zwp_linux_buffer_params_v1_destroy (params);
-      data.wbuf = NULL;
-    }
-  }
+  data.wbuf =
+      zwp_linux_buffer_params_v1_create_immed (params, width, height, fourcc,
+      flags);
+  zwp_linux_buffer_params_v1_destroy (params);
 
 out:
   if (!data.wbuf) {
@@ -192,9 +145,5 @@ out:
         data.wbuf, width, height, GST_FOURCC_ARGS (fourcc), nplanes);
   }
 
-  g_mutex_unlock (&data.lock);
-  g_mutex_clear (&data.lock);
-  g_cond_clear (&data.cond);
-
   return data.wbuf;
 }
-- 
2.20.1

